File: logcal/Models/DietStyle.swift
//
//  DietStyle.swift
//  logcal
//
//  Created by Antigravity on 04/06/26.
//

import Foundation

enum DietStyle: String, CaseIterable, Identifiable {
    case balanced = "Balanced"
    case highProtein = "High Protein"
    case lowCarb = "Low Carb"
    case keto = "Ketogenic"
    case custom = "Custom"
    
    var id: String { self.rawValue }
    
    var macroPercentages: (protein: Double, carbs: Double, fat: Double) {
        switch self {
        case .balanced:
            return (0.30, 0.40, 0.30)
        case .highProtein:
            return (0.40, 0.30, 0.30)
        case .lowCarb:
            return (0.25, 0.15, 0.60)
        case .keto:
            return (0.20, 0.05, 0.75)
        case .custom:
            return (0.30, 0.40, 0.30) // Fallback default
        }
    }
    
    static func calculateGrams(calories: Double, proteinPercent: Double, carbsPercent: Double, fatPercent: Double) -> (protein: Double, carbs: Double, fat: Double) {
        let pGrams = (calories * proteinPercent) / 4.0
        let cGrams = (calories * carbsPercent) / 4.0
        let fGrams = (calories * fatPercent) / 9.0
        
        // Round to nearest integer for clean user display
        return (round(pGrams), round(cGrams), round(fGrams))
    }
}

File: android/app/src/main/java/com/serene/logcal/model/DietStyle.kt
package com.serene.logcal.model

import kotlin.math.round

enum class DietStyle(val rawValue: String) {
    BALANCED("Balanced"),
    HIGH_PROTEIN("High Protein"),
    LOW_CARB("Low Carb"),
    KETO("Ketogenic"),
    CUSTOM("Custom");

    companion object {
        fun fromRawValue(value: String): DietStyle {
            return entries.find { it.rawValue == value } ?: BALANCED
        }

        fun calculateGrams(calories: Double, proteinPercent: Double, carbsPercent: Double, fatPercent: Double): Triple<Double, Double, Double> {
            val pGrams = (calories * proteinPercent) / 4.0
            val cGrams = (calories * carbsPercent) / 4.0
            val fGrams = (calories * fatPercent) / 9.0
            return Triple(round(pGrams), round(cGrams), round(fGrams))
        }
    }

    val macroPercentages: Triple<Double, Double, Double>
        get() = when (this) {
            BALANCED -> Triple(0.30, 0.40, 0.30)
            HIGH_PROTEIN -> Triple(0.40, 0.30, 0.30)
            LOW_CARB -> Triple(0.25, 0.15, 0.60)
            KETO -> Triple(0.20, 0.05, 0.75)
            CUSTOM -> Triple(0.30, 0.40, 0.30)
        }
}
